### PR TITLE
Baw-Client Migration

### LINKED ISSUE #257: Merge with AngularJS Client
Merge this project with the AngularJS Client pages. This should be an Angular10 app which contains an AngularJS app instead of the other way round.

Part of https://github.com/QutEcoacoustics/baw-client/issues/343

Relevant Links: https://medium.com/sv-blog/running-angularjs-1-6-in-angular-5-side-by-side-d2ed771e2a8f

### PR DESCRIPTION
**Pull request by Allcharles**
# Migration of the Baw Client

This is the initial work on implementing the migration of the baw client into the workbench-client. The approach taken was to use iframes, running the baw-client from the assets folder of the workbench-client.

## Changes

- ~Modified app and home components to enable full size page components~
- Created audio analysis pages (mostly a placeholder for testing)
  - `ListComponent`
  - `NewComponent`
- ~Created iframe component (rename to `baw-client`?)~
  - Created `baw-client` component
- Inserted build assets from baw-client into `./src/assets/old-client`
  - Use the [`migration`](https://github.com/QutEcoacoustics/baw-client/tree/migration) branch of baw-client
  - Run `npm run build -- --development`
  - Copy build files inside `./bin` folder to `./src/assets/old-client/`

## Problems

- Some baw-client pages do not resize correctly. This may be improved by modifying more global styles, however it is likely that the styles for individual pages may need to be edited (example new audio analysis page having a horizontal scroll).
- Some visual design problems in the audio analysis list and new pages (these are intentionally left for a separate PR as they are only for testing)
- Need to develop a standardized method of testing components which will use the `IframeComponent`. This will be left for a separate PR
- [x] Tests will not pass until there is a method of mocking the baw-client build artifacts on the VM
- [x] Some tests still need to be implemented such as for the baw interceptor

### Audio Analysis Page

image

### New Audio Analysis Page

image

## Final Checklist

- [x] Assign reviewers if you have permission
- [x] Assign labels if you have permission
- [x] Link issues related to PR
- [x] Ensure project linter is not producing any warnings (`npm run lint`)
- [x] Ensure build is passing on all browsers (`npm run test:all`)
- [x] Ensure CI build is passing
- [x] Ensure docker container is passing ([docs](https://github.com/QutEcoacoustics/workbench-client#docker))

### CODE REVIEW

File: src/app/app.component.scss [hunk 1]
```diff
@@ -1,32 +1,92 @@
...
+        > #secondary,
+        > #action {
+          &.hidden {
+            display: none;
+          }
+        }
+      }
+    }
+  }
 }
```
Comment (Allcharles) [inline src/app/app.component.scss]: This needs commenting to explain what is doing what

File: src/app/app.component.ts [hunk 2]
```diff
@@ -15,15 +14,16 @@ import { AppConfigService } from "./services/app-config/app-config.service";
   selector: "baw-root",
   templateUrl: "./app.component.html",
   styleUrls: ["./app.component.scss"],
+  // Need this so that router-outlet components can be styled
+  // eslint-disable-next-line @angular-eslint/use-component-view-encapsulation
+  encapsulation: ViewEncapsulation.None,
```
Comment (Allcharles) [inline src/app/app.component.ts]: Not sure if this is need anymore. Reverted the changes which affected the sizing of page components

File: src/assets/environment.json [hunk 3]
```diff
@@ -2,7 +2,7 @@
   "environment": {
     "environment": "development",
     "apiRoot": "https://staging.ecosounds.org",
-    "siteRoot": "http://localhost:4200",
```
Comment (Allcharles) [inline src/assets/environment.json:5]: Revert this change? Running this default environment file only works with `npm run start:dev`

**Comment by atruskie** [comment]
I haven't reviewed this yet, but given the tight timelines I'm going to suggest we add the citizen science pages to the list of iframe stubbed pages

File: package.json [hunk 4]
```diff
@@ -13,6 +13,7 @@
     "ng": "node --max_old_space_size=4096 ./node_modules/@angular/cli/bin/ng",
     "nx": "nx",
     "start": "nx serve",
+    "start:dev": "nx serve --host 0.0.0.0 --disableHostCheck",
```
Comment (atruskie) [inline package.json]: why not just change start to do what `start:dev` does?

File: src/app/app.component.html [hunk 5]
```diff
@@ -1,42 +1,23 @@
...
-        <!-- Secondary Menu (only active on menuLayout) -->
-        <div
-          id="secondary"
-          class="p-0 order-1 col-lg-2 col-sm-12"
-          [ngClass]="{ hidden: fullscreen }"
-        >
-          <router-outlet name="secondary"></router-outlet>
-        </div>
+  <div [ngClass]="{ container: !fullscreen, 'container-fluid': fullscreen }">
+    <!-- Secondary Menu (only active on menuLayout) -->
```
Comment (atruskie) [inline src/app/app.component.html]: "only active on menuLayout" - what does this mean?

File: src/app/app.component.html [hunk 6]
```diff
@@ -1,42 +1,23 @@
...
-        <!-- Default and Error Menus -->
-        <div
-          id="page"
-          class="order-lg-2 order-3 col-lg-8 col-12"
-          [ngClass]="{ page: !fullscreen, 'full-width': fullscreen }"
-        >
-          <router-outlet></router-outlet>
-          <router-outlet name="error"></router-outlet>
-        </div>
+    <!-- Default and Error Menus -->
```
Comment (atruskie) [inline src/app/app.component.html]: Is an error really a menu?

File: src/app/app.component.html (hunk 5 above)
Comment (atruskie) [reply src/app/app.component.html]: I'm guessing it means this outlet is used when the menu component is rendered? But given our routing setup, this happens the same time as the main router-outlet rendering, right?

File: src/app/app.component.scss [hunk 7]
```diff
@@ -1,32 +1,114 @@
...
+    }
   }
-}
 
-div.container {
-  min-height: 100%;
-}
+  // Menu Layout
+  > .container {
+    // Compact menus on large displays
```
Comment (atruskie) [inline src/app/app.component.scss]: ? wouldn't we want compact menu on small displays?

File: src/app/app.component.scss [hunk 8]
```diff
@@ -1,32 +1,114 @@
...
+      padding: 0;
+      width: 100%;
+    }
 
-  &:last-child {
-    min-height: 100%;
+    // Hide secondary/action menus when fullscreen
+    > #secondary,> #action {
+      display: none;
+    }
```
Comment (atruskie) [inline src/app/app.component.scss:31]: Confim, when hidden, menus are still accessible, e.g. via clicking a hamburger (menu) icon

File: src/app/app.component.scss (hunk 7 above)
Comment (atruskie) [reply src/app/app.component.scss]: If I understand the `@include` media plugin correctly, it is doing that, just comment is misleading

File: src/app/app.component.ts [hunk 9]
```diff
@@ -40,54 +44,13 @@ export class AppComponent extends withUnsubscribe() implements OnInit {
...
-    // Check if component is a page info component and is not set to fullscreen
-    const component: PageComponent = displayComponent.component as any;
-    if (!component.pageInfo) {
-      this.fullscreen = true;
-    } else {
-      this.fullscreen = !!component.pageInfo.fullscreen;
-    }
+  public updatePageLayout(component: any) {
+    this.fullscreen = !!(component?.constructor?.pageInfo as PageInfo)
+      ?.fullscreen;
```
Comment (atruskie) [inline src/app/app.component.ts:59]: Since the mechanics of this mostly happen in templates, add a comment explaining workflow...

When component is loaded, uses its knowledge of pageInfo, to call this method (to escape encapsualtion) and change the fullscreen property, which is bound ??by/from/to what??

File: src/app/components/audio-analysis/pages/list/list.component.spec.ts [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+  beforeEach(() => {
+    fixture = TestBed.createComponent(ListComponent);
+    component = fixture.componentInstance;
+    fixture.detectChanges();
+  });
+
+  it("should create", () => {
+    expect(component).toBeTruthy();
+  });
+});
```
Comment (atruskie) [inline src/app/components/audio-analysis/pages/list/list.component.spec.ts]: pointless test?

if it's just testing costruction, at least make the `toBeTruthy` expectation more useful. Why not test that component is an instance of AudioAnalysis.ListComponent ?

Comment (atruskie) [reply src/app/components/audio-analysis/pages/list/list.component.spec.ts]: Another good test for these pages is asserting the `baw-client` element/component is exists/rendered in the template

File: src/app/components/audio-analysis/pages/new/new.component.spec.ts [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+  });
+
+  beforeEach(() => {
+    fixture = TestBed.createComponent(NewComponent);
+    component = fixture.componentInstance;
+    fixture.detectChanges();
+  });
+
+  it("should create", () => {
+    expect(component).toBeTruthy();
```
Comment (atruskie) [inline src/app/components/audio-analysis/pages/new/new.component.spec.ts]: same comments on tests as with list.component.spec.ts

File: src/app/components/shared/baw-client/baw-client.component.ts [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+        width: 100%;
+        min-height: 100%;
+        border: 0px;
+      }
+    `,
+  ],
+})
+export class BawClientComponent extends withUnsubscribe() implements OnInit {
+  @ViewChild("content") private iframeRef: ElementRef<HTMLIFrameElement>;
+  public url: SafeResourceUrl;
```
Comment (atruskie) [inline src/app/components/shared/baw-client/baw-client.component.ts:42]: this should probably be an`@input` so we can customise what baw-client page is loaded

File: src/assets/environment.json (hunk 3 above)
Comment (atruskie) [reply src/assets/environment.json:5]: no this is a fine change - i've already suggested dev should bind to everything

maybe add a note to the readme that alerts the user that the dev server binds to all interfaces

**Review by atruskie — APPROVED** [review]
There's no details page for the audio-analyses?

> Some baw-client pages do not resize correctly. This may be improved by modifying more global styles, however it is likely that the styles for individual pages may need to be edited (example new audio analysis page having a horizontal scroll).

Ignore. Immediate priority is deploying, and then we'll be rebuilding old pages

> Some visual design problems in the audio analysis list and new pages (these are intentionally left for a separate PR as they are only for testing)

Probably ignore. Depends what the issue is.

> Need to develop a standardized method of testing components which will use the IframeComponent. This will be left for a separate PR

The only tests I want to see are ones that verify:

- the component loads
- the iframe loads
- there's non-error content in the iframe
- 
But little else

File: src/app/components/audio-analysis/pages/list/list.component.spec.ts (hunk 10 above)
Comment (Allcharles) [reply src/app/components/audio-analysis/pages/list/list.component.spec.ts]: This test is disabled, (`xdescribe`). It is implemented later on, this is just the default test generated by angular schematics

File: package.json (hunk 4 above)
Comment (Allcharles) [reply package.json]: Mostly because its a potential security vulnerability for the running computer:
```
Warning: This is a simple server for use in testing or debugging Angular applications
locally. It hasn't been reviewed for security issues.

Binding this server to an open connection can result in compromising your application or
computer. Using a different host than the one passed to the "--host" flag might result in
websocket connection issues. You might need to use "--disableHostCheck" if that's the
case.
Warning: Running a server with --disable-host-check is a security risk. See https://medium.com/webpack/webpack-dev-server-middleware-security-issues-1489d950874a for more information.
```

File: src/app/app.component.html (hunk 5 above)
Comment (Allcharles) [reply src/app/app.component.html]: This is a mistake. A remnant of previous behavior where visibility was triggered by the action css class

File: package.json (hunk 4 above)
Comment (atruskie) [reply package.json]: Yeah, not really. The server is not meant to run in production. That's what they're discouraging 

Basically, if you're running any firewall it is totally fine. Even if not, then, it is likely fine.

Anything running from inside docker has to bind to any address as well because the containers are on different networks from the host and you wouldn't be able to use your dev service otherwise.

File: src/app/app.component.scss (hunk 8 above)
Comment (Allcharles) [reply src/app/app.component.scss:31]: When hidden, menus are not accessible. Turning them into a hamburger menu is a future upgrade to the design of the website. This is tracked by: #458
